openedx_tagging/core/tagging/models.py

```diff
@@ -627,3 +649,240 @@ def copy(self, object_tag: "ObjectTag") -> "ObjectTag":
+        """
+        return (
+            super()._check_taxonomy(object_tag)
+            and self.system_defined
+        )
+
+
+class ModelSystemDefinedTaxonomy(SystemDefinedTaxonomy):
+    """
+    Model based system taxonomy abstract class.
+    
+    This type of taxonomies has an associated Django model in `tag_class_model()`.
+    Are designed to create a Tag when an ObjectTags with a new Tag. 
+    The Tags are representations of the instances of the associated model.
+
+    On Tag.external_id stores an identifier from the instance (`pk` as default)
+    and on Tag.value stores an human readable representation of the instance
+    (e.g. `username`). 
+    The subclasses can override this behavior, to choose the right field.
+
+    When an ObjectTag is created with an existing Tag, 
+    the Tag is re-synchronized with its instance.
+    """
+
+    class Meta:
+        proxy = True
+
+    def __init__(
+        self,
+        *args: Any,
+        **kwargs: Any
+    ) -> None:
+        """
+        Checks if the `tag_class_model` is correct
+        """
+        assert issubclass(self.tag_class_model, models.Model)
+        super().__init__(*args, **kwargs)
+
+    @property
+    def tag_class_model(self) -> Type:
+        """
+        Subclasses must implement this method to return the Django.model
+        class referenced by these object tags.
+        """
+        raise NotImplementedError
+    
+    def _check_instance(
+        self,
+        object_tag: ObjectTag
+    ) -> bool:
+        """
+        Returns True if the instance exists
+
+        Subclasses can override this method to perform their own instance validation checks.
+        """
+        intance_id = object_tag.tag.external_id
+        return self.tag_class_model.objects.filter(pk=intance_id).exists()
+
+    def _check_tag(
+        self,
+        object_tag: ObjectTag
+    ) -> bool:
+        """
+        Returns True if pass the instance validation checks
+        """
+        return (
+            super()._check_tag(object_tag)
+            and self._check_instance(object_tag)
+        )
+    
+    def _get_external_id_from_instance(
+        self,
+        instance
+    ) -> str:
+        """
+        Returns the tag external id from the instance
+
+        Subclasses can override this method to get the external id from their own models.
+        """
+        return str(instance.pk)
+
+    def _get_value_from_instance(
+        self,
+        instance
+    ) -> str:
+        """
+        Returns the tag value from the instance
+
+        Subclasses can override this method to get the value from their own models.
+        """
+        return str(instance.pk)
+    
+    def _get_instance(self, pk):
+        """
+        Gets the instance from `pk`
+        """
+        try:
+            return self.tag_class_model.objects.get(pk=pk)
+        except self.tag_class_model.DoesNotExist:
+            return None
+    
+    def _resync_tag(self, tag: Tag) -> Tag:
```

Comment:
I think it's cleaner to define a `ModelObjectTag` subclass of ObjectTag to handle most of this functionality, cf https://github.com/open-craft/openedx-learning/pull/6/commits/9fd8854c293cc9b506c2a55181c5aab9f6b2b80d

Reply:
Added to the branch